### PR TITLE
[Skia] Fix support for custom cursors

### PR DESCRIPTION
**Pull request by TingPing**
#### 1853a763ab9961b63d400a20f9f04f64eb2bf184
<pre>
[Skia] Fix support for custom cursors
<a href="https://bugs.webkit.org/show_bug.cgi?id=270389">https://bugs.webkit.org/show_bug.cgi?id=270389</a>

Reviewed by Carlos Garcia Campos.

* Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp:
(WebCore::ShareableBitmap::createImage):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::setCursor):

Canonical link: <a href="https://commits.webkit.org/275665@main">https://commits.webkit.org/275665@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9916b2e775a5e8228c76c22288ca446bb18230bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42336 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44935 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38453 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44643 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18698 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35077 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16007 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15999 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46397 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37835 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41744 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17152 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14136 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40343 "Found 1 new API test failure: /WebKitGTK/TestWebKitFaviconDatabase:/webkit/WebKitFaviconDatabase/get-favicon (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18771 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18833 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5727 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18416 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp [hunk 1]
```diff
@@ -69,8 +69,8 @@ void ShareableBitmap::paint(GraphicsContext&, float /*scaleFactor*/, const IntPo
 
 RefPtr<Image> ShareableBitmap::createImage()
 {
-    notImplemented();
-    return nullptr;
+    auto image = SkImages::RasterFromData(m_configuration.imageInfo(), SkData::MakeWithCopy(data(), sizeInBytes()), bytesPerRow());
```
Comment (TingPing) [inline Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp:72]: I don't think Skia has any equivalent of `cairo_surface_set_user_data()` so this is yet another copy of data...

File: Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp [hunk 2]
```diff
@@ -69,14 +69,24 @@ void ShareableBitmap::paint(GraphicsContext&, float /*scaleFactor*/, const IntPo
...
+}
+
+PlatformImagePtr ShareableBitmap::createPlatformImage(BackingStoreCopy backingStoreCopy, ShouldInterpolate)
+{
+    sk_sp<SkData> pixelData;
+    if (backingStoreCopy == BackingStoreCopy::CopyBackingStore)
+        pixelData = SkData::MakeWithCopy(data(), sizeInBytes());
+    else {
+        pixelData = SkData::MakeWithProc(data(), sizeInBytes(), releasedSkDataRef, this);
+        ref(); // Balanced by unref in releasedSkDataRef.
```
Comment (carlosgcampos) [inline Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp]: I would call ref() before and use a lambda for the release.

File: Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp [hunk 3]
```diff
@@ -952,7 +958,21 @@ void View::setCursor(const WebCore::Cursor& cursor)
...
+        return;
+
+    SkPixmap pixmap;
+    auto platformImage = nativeImage->platformImage();
+    ASSERT(platformImage->peekPixels(&pixmap));
+
+    GRefPtr<GBytes> bytes = adoptGRef(g_bytes_new_with_free_func(pixmap.addr(), pixmap.computeByteSize(), [](gpointer data) {
+        static_cast<SkImage*>(data)->unref();
+    }, platformImage.get()));
+    platformImage->ref();
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp]: Same here, I would call ref before. It's not very important and probably just a matter of taste, but visually you see the ref first and then the unref. In any case, don't change it if you like this way more.

**Review by carlosgcampos — APPROVED** [review]
Thanks!

File: Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp (hunk 3 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp]: Another option here is calling `SkRef` that calls `ref()` and returns the pointer so you can pass it directly to `g_bytes_new_with_free_func`

**Pull request merged**